### ISSUE TITLE
Wrong indexing for broadcasting before reduction

### ISSUE THREAD

**Issue by jjsjann123** [csarofeen/pytorch#456]
## 🐛 Bug

A common broadcast + reduction pattern is broken on ToT. I have pushed a cpp repro to: https://github.com/csarofeen/pytorch/commit/3e510ddad6a23e732cf3136a56f152217185baf9

~The issue was introduced in a recent PR: https://github.com/csarofeen/pytorch/pull/423 (commit: 9e20a2d6b115df98db917d41ee7017fa55577696)~
This is not a regression, I was stupid and made terrible mistakes during bisecting. It is surprising that we have issue in this simple case, I should re-visit our python tests to expand the coverage on broadcast + reduction.

cpp example:
```
TEST(NVFuserTest, FusionBroadcastReductionRepro_CUDA) {
  Fusion fusion;
  FusionGuard fg(&fusion);

  std::vector<int64_t> tensor0_shape{7, 4, 7};
  std::vector<int64_t> tensor1_shape{4, 7};

  TensorView* tv0 = makeDummyTensor(tensor0_shape.size());
  fusion.addInput(tv0);
  TensorView* tv1 = makeDummyTensor(tensor1_shape.size());
  fusion.addInput(tv1);

  TensorView* tv2 = add(tv0, tv1);
  TensorView* tv3 = reductionOp(BinaryOpType::Add, {0, 1}, new Float(0), tv2);
  fusion.addOutput(tv3);

  const auto options =
      at::TensorOptions().dtype(at::kFloat).device(at::kCUDA, 0);

  at::Tensor input0 = at::randn(tensor0_shape, options);
  at::Tensor input1 = at::randn(tensor1_shape, options);

  std::vector<int64_t> reduction_axes{0, 1};
  auto reduction_params = getReductionHeuristics(&fusion, {input0, input1}, tv3);
  TORCH_CHECK(reduction_params, "Reduction schedule was not generated!");
  scheduleReduction(&fusion, reduction_params.value(), tv3, {});
  FusionExecutor fe;
  fe.compileFusion(&fusion);
  auto outputs = fe.runFusion({input0, input1}, reduction_params.value().lparams);

  auto aten_output = input0.add(input1).sum(reduction_axes);

  TORCH_CHECK(
      aten_output.allclose(outputs[0], 1e-04, 1e-04),
      "Error of: ",
      aten_output.sub(outputs[0]).abs().max());
}
```

This is the generated code, looking at the very beginning, indexing into `T1` looks very wrong:
```
__global__ void kernel1(Tensor<float, 3> T0, Tensor<float, 2> T1, Tensor<float, 1> T4) {
  alignas(4) extern __shared__ char array[];
  void* shared_mem = array;
  if ((((blockIdx.x * blockDim.x) + threadIdx.x) < T0.size[2])) {
    T4[(((blockIdx.x * blockDim.x) + threadIdx.x) * T4.stride[0])] = float(0);
  }
  float T5[1];
  if ((((blockIdx.x * blockDim.x) + threadIdx.x) < T0.size[2])) {
    T5[0] = float(0);
  }
  for(size_t ki41 = 0; ki41 < (ceilDiv((ceilDiv((T0.size[0] * T0.size[1]), blockDim.y)), 4)); ++ki41) {
    float T2[4];
    // Alias Allocation - register
    float* T3 = T2;
    if ((((((blockIdx.x * blockDim.x) + threadIdx.x) < T0.size[2]) && ((((((ki41 * 4) + (4 - 1)) * blockDim.y) + threadIdx.y) / T0.size[1]) < T0.size[0])) && ((((((ki41 * 4) + (4 - 1)) * blockDim.y) + threadIdx.y) % T0.size[1]) < T0.size[1]))) {
      for(size_t ki44 = 0; ki44 < 4; ++ki44) {
        T2[ki44]
           = T1[(((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) * T1.stride[0]) + (((blockIdx.x * blockDim.x) + threadIdx.x) * T1.stride[1])];
        T3[ki44]
          = T0[((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) / T0.size[1]) * T0.stride[0]) + ((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) % T0.size[1]) * T0.stride[1]) + (((blockIdx.x * blockDim.x) + threadIdx.x) * T0.stride[2])]
          + T2[ki44];
        T5[0]
          = T5[0]
          + T3[ki44];
      }   
    } else {
      for(size_t ki44 = 0; ki44 < 4; ++ki44) {
        if (((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) < T1.size[0]) && (((blockIdx.x * blockDim.x) + threadIdx.x) < T1.size[1]))) {
          T2[ki44]
             = T1[(((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) * T1.stride[0]) + (((blockIdx.x * blockDim.x) + threadIdx.x) * T1.stride[1])];
        }
        if (((((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) / T0.size[1]) < T0.size[0]) && ((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) % T0.size[1]) < T0.size[1])) && (((blockIdx.x * blockDim.x) + threadIdx.x) < T0.size[2]))) {
          T3[ki44]
            = T0[((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) / T0.size[1]) * T0.stride[0]) + ((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) % T0.size[1]) * T0.stride[1]) + (((blockIdx.x * blockDim.x) + threadIdx.x) * T0.stride[2])]
            + T2[ki44];
        }   
        if (((((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) / T0.size[1]) < T0.size[0]) && ((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) % T0.size[1]) < T0.size[1])) && (((blockIdx.x * blockDim.x) + threadIdx.x) < T0.size[2]))) {
          T5[0]
            = T5[0]
            + T3[ki44];
        }   
      } 
    } 
  }
  blockReduce<false, true, false>(  
    T4[(((blockIdx.x * blockDim.x) + threadIdx.x) * T4.stride[0])],
    T5[0],
    [](float &a, float b) { a = a + b; },
    threadIdx,
    blockDim,
    static_cast<float*>(shared_mem),
    (((blockIdx.x * blockDim.x) + threadIdx.x) < T0.size[2]),
    float(0));
}
```

**Comment by jjsjann123** [comment]
Apology to @naoyam for pointing at your innocent PR earlier. Turns out this is not a regression, but a new issue that we should further investigate.

**Comment by jjsjann123** [comment]
```
        T2[ki44]
           = T1[(((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) * T1.stride[0]) + (((blockIdx.x * blockDim.x) + threadIdx.x) * T1.stride[1])];
        T3[ki44]
          = T0[((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) / T0.size[1]) * T0.stride[0]) + ((((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) % T0.size[1]) * T0.stride[1]) + (((blockIdx.x * blockDim.x) + threadIdx.x) * T0.stride[2])]
          + T2[ki44];
```

T0 with shape {7, 4, 7} and T1 with shape {4, 7}
If we look at this section, indexing into `T0` looks fine, but `T1` should be broadcasted on `T0.axis(0)`, so its indexing should be
```
T1[(((((ki41 * 4) + ki44) * blockDim.y) + threadIdx.y) % T0.size[1] * T1.stride[0]) + (((blockIdx.x * blockDim.x) + threadIdx.x) * T1.stride[1])]
```

**Comment by naoyam** [comment]
The updated reproducer is here: https://github.com/csarofeen/pytorch/blob/bcast_reduction_repro/test/cpp/jit/test_gpu.cpp#L5642

**Comment by naoyam** [comment]
Closing as this should be fixed by #471.
